File: 2018/day9/python/main.py
class Node(object):
    def __init__(self, v):
        self.value = v
        self.next = None
        self.prev = None

totalPlayers = 418
lastMarble = 70769
superLastMarble = lastMarble*100

def play(end):
    currPlayer = 0
    score = [0]*totalPlayers
    currMarble = Node(0)
    currMarble.next = currMarble
    currMarble.prev = currMarble
    for marble in range(1, end+1):
        if marble % 23 != 0:
            newMarble = Node(marble)
            currMarble = currMarble.next
            newMarble.prev = currMarble
            newMarble.next = currMarble.next
            currMarble.next.prev = newMarble
            currMarble.next = newMarble
            currMarble = newMarble
        else:
            score[currPlayer] += marble
            currMarble = currMarble.prev.prev.prev.prev.prev.prev.prev
            score[currPlayer] += currMarble.value
            currMarble = currMarble.next
            currMarble.prev.prev.next = currMarble
            currMarble.prev = currMarble.prev.prev
        currPlayer = (currPlayer + 1) % totalPlayers
    return max(score)

print("high score: %d" % play(lastMarble))
print("high score: %d" % play(superLastMarble))

File: 2018/day9/go/src/main.go
package main

import (
	"fmt"
)

type node struct {
	Value int
	Next  *node
	Prev  *node
}

func max(array []int) (value int) {
	for _, v := range array {
		if v > value {
			value = v
		}
	}
	return
}

func play(numPlayers int, lastMarble int) int {
	score := make([]int, numPlayers)
	currentPlayer := 0
	currentMarble := &node{Value: 0}
	currentMarble.Next = currentMarble
	currentMarble.Prev = currentMarble
	for marble := 1; marble < lastMarble; marble++ {
		if marble%23 != 0 {
			newMarble := &node{Value: marble}
			currentMarble = currentMarble.Next
			newMarble.Prev = currentMarble
			newMarble.Next = currentMarble.Next
			currentMarble.Next.Prev = newMarble
			currentMarble.Next = newMarble
			currentMarble = newMarble
		} else {
			score[currentPlayer] += marble
			currentMarble = currentMarble.Prev.Prev.Prev.Prev.Prev.Prev.Prev
			score[currentPlayer] += currentMarble.Value
			currentMarble = currentMarble.Next
			currentMarble.Prev.Prev.Next = currentMarble
			currentMarble.Prev = currentMarble.Prev.Prev
		}
		currentPlayer = (currentPlayer + 1) % numPlayers
	}
	return max(score)
}

func main() {
	const numPlayers = 418
	const lastMarble = 70769
	const lastMarble2 = lastMarble * 100

	fmt.Printf("High score #1: %d\n", play(numPlayers, lastMarble))
	fmt.Printf("High score #2: %d\n", play(numPlayers, lastMarble2))
}
